### PR TITLE
BUG: optimize: avoid an infinite loop in Newton-CG

### LINKED ISSUE #6165: optimize.minimize infinite loop with Newton-CG
I am currently trying to use Newton-CG but am running into a problem where I encounter an infinite loop in the minimization routine. Code to replicate this may be found here:

https://gist.github.com/Chris7/51ed3a8f8cec011ce3342615675195b7

I am using Ubuntu 14.04 64 bit, and here is my pip freeze:
```
coverage==4.0.3
coveralls==1.1
Cython==0.23.4
docopt==0.6.2
line-profiler==1.0
lxml==3.5.0
memory-profiler==0.41
nose==1.3.7
numpy==1.10.1
pandas==0.17.1
patsy==0.4.1
plotly==1.9.6
pluggy==0.3.1
profilestats==2.0
py==1.4.31
pyprof2calltree==1.3.2
python-dateutil==2.4.2
pytz==2015.7
requests==2.8.1
scikit-learn==0.17
scipy==0.17.1
six==1.10.0
tox==2.2.1
virtualenv==13.1.2
```

The code which runs infinitely is in optimize.optimize, lines 1412-1440 and are:

```python
while numpy.add.reduce(numpy.abs(ri)) > termcond:
    if fhess is None:
        if fhess_p is None:
            Ap = approx_fhess_p(xk, psupi, fprime, epsilon)
        else:
            Ap = fhess_p(xk, psupi, *args)
            hcalls = hcalls + 1
    else:
        Ap = numpy.dot(A, psupi)
    # check curvature
    Ap = asarray(Ap).squeeze()  # get rid of matrices...
    curv = numpy.dot(psupi, Ap)
    if 0 <= curv <= 3 * float64eps:
        break
    elif curv < 0:
        if (i > 0):
            break
        else:
            # fall back to steepest descent direction
            xsupi = dri0 / (-curv) * b
            break
    alphai = dri0 / curv
    xsupi = xsupi + alphai * psupi
    ri = ri + alphai * Ap
    dri1 = numpy.dot(ri, ri)
    betai = dri1 / dri0
    psupi = -ri + betai * psupi
    i = i + 1
    dri0 = dri1          #
```

I would try to patch this myself, but I do not know enough about the Newton-CG method (yet). The  issue appears to be that xsupi never increments and the curvature keeps increasing so it never breaks.

### PR DESCRIPTION
**Pull request by ev-br**
Mechanically cap the number of iterations in the inner loop.

closes gh-6165

### CODE REVIEW

File: scipy/optimize/optimize.py [hunk 1]
```diff
@@ -1465,7 +1465,8 @@ def _minimize_newtoncg(fun, x0, args=(), jac=None, hess=None, hessp=None,
             A = fhess(*(xk,) + args)
             hcalls = hcalls + 1
 
-        while numpy.add.reduce(numpy.abs(ri)) > termcond:
+        k2 = 0
```
Comment (nmayorov) [inline scipy/optimize/optimize.py]: I realize that there are probably many places to improve clarity in this code. Buy maybe name `k2` as `cg_iter` or something like this.

File: scipy/optimize/optimize.py [hunk 2]
```diff
@@ -1465,7 +1465,8 @@ def _minimize_newtoncg(fun, x0, args=(), jac=None, hess=None, hessp=None,
...
+        while (numpy.add.reduce(numpy.abs(ri)) > termcond) and (k2 < maxiter):
```
Comment (nmayorov) [inline scipy/optimize/optimize.py]: `maxiter` is a not good limiting value here. CG iterations converge in at most `number of variables` iterations, let's put that value here.

Let's remove parentheses.

File: scipy/optimize/optimize.py [hunk 3]
```diff
@@ -1535,6 +1541,16 @@ def _minimize_newtoncg(fun, x0, args=(), jac=None, hess=None, hessp=None,
             print("         Function evaluations: %d" % fcalls[0])
             print("         Gradient evaluations: %d" % gcalls[0])
             print("         Hessian evaluations: %d" % hcalls)
+    elif k2 >= maxiter:
+        warnflag = 3
+        msg = "Hessian computation infinite loop."
```
Comment (nmayorov) [inline scipy/optimize/optimize.py]: Rather "CG iterations didn't converge".

**Comment by dlax** [comment]
Wouldn't a `for` loop on `range(maxiter)` with a `break` on termination condition be clearer?

File: scipy/optimize/optimize.py (hunk 2 above)
Comment (ev-br) [inline scipy/optimize/optimize.py]: If it converges in order-N iterations *under reasonably weak conditions*, then I'd say all we need to do is to pick a multiple of N. Like, 10N or 50N or maybe keep it

```
    if maxiter is None:
        maxiter = len(x0)*200
```
?

**Comment by ev-br** [comment]
The style of this loop and the variable `k2` was taken to mimic the outer loop, https://github.com/scipy/scipy/blob/master/scipy/optimize/optimize.py#L1451

```
    while (numpy.add.reduce(numpy.abs(update)) > xtol) and (k < maxiter):
```

So I just went for an absolute minimal tweak. Do you want me to fix both loops?

File: scipy/optimize/optimize.py (hunk 2 above)
Comment (nmayorov) [inline scipy/optimize/optimize.py]: As far as I remember the classical CG method is applicable if and only if the matrix is positive defined. If it does converge, then in n iterations: https://en.wikipedia.org/wiki/Conjugate_gradient_method

So the classical Newton-CG is applicable for convex problems only, when the Hessian is positive definite.

My suggestion is to change ``k2 < len(x0)``.

Comment (nmayorov) [inline scipy/optimize/optimize.py]: So, yes I think it's correct to limit it to `n` iterations. If it does `n` iterations, then check the tolerance and if it is yet not satisfied, report that the CG iterations didn't converge. (Alternatively run `n+1` iterations and if `n+1` is done, then the CG iterations didn't converge).

The message probably could be: "CG iteration didn't converge, the Hessian is not positive definite." (Sorry, forgot to mention Hessian in the initial suggestion).

**Comment by dlax** [comment]
Evgeni Burovski a écrit :
> The style of this loop and the variable |k2| was taken to mimic the
> outer loop,
> https://github.com/scipy/scipy/blob/master/scipy/optimize/optimize.py#L1451
>
> |while (numpy.add.reduce(numpy.abs(update)) > xtol) and (k < maxiter): |
>
> So I just went for an absolute minimal tweak. Do you want me to fix both
> loops?
>

Oh, I overlooked this outer loop. But yes, I think this would be clearer 
with `for` loops at both places. You could then have the outer one break 
in the else clause of the inner one, and also handle formating the `msg` 
and prints there (this currently comes later on with a redundant `if k2 
 >= maxiter`).

File: scipy/optimize/optimize.py (hunk 2 above)
Comment (nmayorov) [inline scipy/optimize/optimize.py]: I found MATLAB's defaults for `maxit` in CG: http://www.mathworks.com/help/matlab/ref/pcg.html, so it's `min(20, n)`. Can't really comment about this 20, should be some reason for that.

**Comment by ev-br** [comment]
> You could then have the outer one break
> in the else clause of the inner one, and also handle formating the `msg`
> and prints there (this currently comes later on with a redundant `if k2
> = maxiter`).

Maybe I'm just dense, I don't understand how that would work. 

Would you guys agree on fixing the bug now and addressing the style issues separately?

File: scipy/optimize/optimize.py [hunk 4]
```diff
@@ -1535,6 +1541,16 @@ def _minimize_newtoncg(fun, x0, args=(), jac=None, hess=None, hessp=None,
             print("         Function evaluations: %d" % fcalls[0])
             print("         Gradient evaluations: %d" % gcalls[0])
             print("         Hessian evaluations: %d" % hcalls)
+    elif k2 >= maxiter:
+        warnflag = 3
+        msg = "Warning: CG iterations didn't converge."
```
Comment (Chris7) [inline scipy/optimize/optimize.py]: I would add a comment that the hessian may be incorrect. Mine actually was and this bug helped me track the error :)

**Comment by dlax** [comment]
Evgeni Burovski a écrit :
>     You could then have the outer one break
>     in the else clause of the inner one, and also handle formating the |msg|
>     and prints there (this currently comes later on with a redundant |if k2
>     = maxiter|).
>
> Maybe I'm just dense, I don't understand how that would work.
>
> Would you guys agree on fixing the bug now and addressing the style
> issues separately?

This is fine for me. I have something on top of your branch ready; will 
submit a PR once that one is merged.

@nmayorov is this ready?

**Comment by nmayorov** [comment]
@dlax I think that `k2` should be compared with something proportional to `n`. As another reference `minimize(method='TNC')` (actually it is the Newton-CG most likely decently implemented, do we really need `Newton-CG` duplicate?) uses:

```
maxCGit : int
Maximum number of hessian*vector evaluations per main iteration. If maxCGit == 0, 
the direction chosen is -gradient if maxCGit < 0, maxCGit is set to max(1,min(50,n/2)).
Defaults to -1.
```

So we can adopt `min(50, n/2)`, if I were implementing the method I would put simply `n`.

**Comment by nmayorov** [comment]
As a minimal fix it's fine: if the method failed - it failed once, doesn't really matter how many iterations were done in CG loop. Anyway it needs some clean-up (or maybe better deprecation, consider 'TNC').

>I would add a comment that the hessian may be incorrect. Mine actually was and this bug helped me track the error :)

This can be helpful to address: "CG iteration didn't converge, the Hessian is not positive definite."

**Comment by ev-br** [comment]
Updated with a better warning message and `cg_maxiter = 20*len(x)`.

**Pull request merged**